dashboard/config/libraries/PoemBot.interpreted.js

```diff
@@ -1,49 +1,119 @@
+// Poem: An object that has field for title, author, and lines. The title and author
+// fields should both have type Text. The lines field should be an array of Line values.
+
+// CONSTANTS
+var outerMargin = 50;
+var defaultLineHeight = 50;
+var defaultFontSize = 25;
+var blankText = {value: '', type: 'blank'};
+
+// STATE
+var poem = {
+  title: blankText,
+  author: blankText,
+  lines: []
+};
+var isVisible = false;
+var backgroundEffect = function() {};
+var foregroundEffect = function() {};
+
+
+// This function runs as part of the draw loop
 function poemBotDraw() {
+  backgroundEffect();
+  foregroundEffect();
   if (isVisible) {
     drawPoem();
   }
 }
+// Push our PoemBot draw function into the NativeSpriteLab.interpreted.js
+// draw loop. See NativeSpriteLab.interpreted.js for usage.
+other.push(poemBotDraw);
 
 function drawPoem() {
-  var outerMargin = 50;
-  var defaultLineHeight = 50;
-  var defaultFontSize = 25;
-  var cursor = outerMargin;
+  var yCursor = outerMargin;
   fill('black');
   noStroke();
   textSize(defaultFontSize);
   textAlign(CENTER);
   if (poem.title) {
     textSize(defaultFontSize * 2);
-    text(poem.title, 200, cursor);
+    text(poem.title, 200, yCursor);
     textSize(defaultFontSize);
-    cursor += defaultLineHeight;
+    yCursor += defaultLineHeight;
   }
   if (poem.author) {
-    cursor -= defaultLineHeight / 2;
+    yCursor -= defaultLineHeight / 2;
     textSize(16);
-    text("by: " + poem.author, 200, cursor);
+    text(poem.author, 200, yCursor);
     textSize(defaultFontSize);
-    cursor += defaultLineHeight;
+    yCursor += defaultLineHeight;
   }
 
-  var lineHeight = (400 - cursor) / poem.lines.length;
+  var lineHeight = (400 - yCursor) / poem.lines.length;
   for (var i = 0; i < poem.lines.length; i++) {
-    var line = poem.lines[i];
-    var width = textWidth(line);
-    textSize(Math.min(defaultFontSize, defaultFontSize * (400 - outerMargin) / width));
-    text(line, 200, cursor);
+    drawPoemLine(poem.lines[i], yCursor);
     textSize(defaultFontSize);
-    cursor += lineHeight;
+    yCursor += lineHeight;
   }
 }
 
-var poem = {
-  title: undefined,
-  author: undefined,
-  lines: []
-};
-var isVisible = false;
+function drawPoemLine(line, yPos) {
+  // Concatenate all the text values together so we can compute the length
+  // of the printed text
+  var words = [];
+  var i = 0;
+  for (i = 0; i < line.length; i++) {
+    words.push(line[i].value);
+  }
+  var fullLine = words.join(' ');
+  var fullWidth = textWidth(fullLine);
+  textSize(Math.min(defaultFontSize, defaultFontSize * (400 - outerMargin) / fullWidth));
+
+  // recompute with scaled textSize
+  fullWidth = textWidth(fullLine);
+
+  var start = 200 - fullWidth/2;
+  textAlign(LEFT);
+  var xCursor = start;
+  for (i = 0; i < line.length; i++) {
+    // TODO: Make font colors configurable by students
+    if (line[i].type == "random") {
+      fill("blue");
+    } else {
+      fill("black");
+    }
```

Comment:
just making sure i understand our intent here -- we'll fill random words in a different color than any other type of word, and we will allow them to configure the color (e.g., `set _random_ word to _purple_`) of each type?

Reply:
Yes, that's correct (at least for now, I think curriculum is still iterating here)